File: common/web/src/main/java/com/example/chat/common/web/response/ApiResponse.java
package com.example.chat.common.web.response;

import com.fasterxml.jackson.annotation.JsonInclude;
import org.springframework.http.ResponseEntity;

/**
 * 공통 API 응답 DTO
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record ApiResponse<T>(
        int status,
        String message,
        T data) {

    public static <T> ApiResponse<T> ok(T data) {
        return new ApiResponse<>(200, "OK", data);
    }

    public static ApiResponse<Void> ok() {
        return new ApiResponse<>(200, "OK", null);
    }

    public static <T> ApiResponse<T> created(T data) {
        return new ApiResponse<>(201, "Created", data);
    }

    public static ApiResponse<Void> created() {
        return new ApiResponse<>(201, "Created", null);
    }

    public static ApiResponse<Void> noContent() {
        return new ApiResponse<>(204, "No Content", null);
    }

    public ResponseEntity<ApiResponse<T>> toResponseEntity() {
        return ResponseEntity.status(status).body(this);
    }
}

File: common/web/src/main/java/com/example/chat/common/web/response/ApiResponse.kt
package com.example.chat.common.web.response

import com.fasterxml.jackson.annotation.JsonInclude
import org.springframework.http.ResponseEntity

@JsonInclude(JsonInclude.Include.NON_NULL)
data class ApiResponse<T>(
	val status: Int,
	val message: String?,
	val data: T?
) {
	companion object {
		fun <T> ok(data: T?) =
			ApiResponse<T?>(status = 200, message = "OK", data = data)

		fun ok(): ApiResponse<Void?> =
			ok<Void?>(null)

		fun <T> created(data: T?) =
			ApiResponse<T?>(status = 201, message = "Created", data = data)

		fun created(): ApiResponse<Void?> =
			created<Void?>(null)

		fun noContent() =
			 ApiResponse<Void?>(status = 204, message = "No Content", data = null)
	}

	fun toResponseEntity(): ResponseEntity<ApiResponse<T>> =
		ResponseEntity.status(this.status).body(this)
}
